lib/translator/read.cc

```diff
@@ -0,0 +1,181 @@
+#include "read.h"
+
+#include <cmath>
+
+#include "absl/base/casts.h"
+#include "absl/types/span.h"
+#include "common.h"
+
+namespace translator {
+
+namespace {  // anonymous namespace for helper functions
+
+uint8_t GetPrinfo(uint8_t rd_protect) {
+  bool pract;      // Protection Information Action 1 bit
+  uint8_t prchk;   // Protection Information Check 3 bits
+  uint8_t prinfo = 0;  // 4 bits
+
+  switch (rd_protect) {
+    case 0b0:
+      pract = 0b1;
+      prchk = 0b111;
+      break;
+    case 0b001:
+    case 0b101:
+      pract = 0b0;
+      prchk = 0b111;
+      break;
+    case 0b010:
+      pract = 0b0;
+      prchk = 0b011;
+      break;
+    case 0b011:
+      pract = 0b0;
+      prchk = 0b000;
+      break;
+    case 0b100:
+      pract = 0b0;
+      prchk = 0b100;
+      break;
+    default:
+      // shouldn't reach here because we covered all possible values of
+      // rd_protect
+      break;
+  }
+
+  prinfo |= prchk;
+  prinfo |= (pract << 3);
+
+  return prinfo;
+}
+
+void SetLbaTags(uint32_t eilbrt, uint16_t elbat, uint16_t elbatm,
+                nvme_defs::GenericQueueEntryCmd& queue_entry) {
+  // expected initial logical block reference
+  queue_entry.cdw[4] = eilbrt;
+  // expected logical block application tag
+  queue_entry.cdw[5] |= elbat;
+  // bits  31:16 expected logical block application tag mask
+  queue_entry.cdw[5] |= (elbatm << 16);
+}
+
+void LegacyRead(uint32_t nsid, uint32_t lba, uint16_t transfer_length,
+                nvme_defs::GenericQueueEntryCmd& queue_entry) {
+  queue_entry.opc = static_cast<uint8_t>(nvme_defs::NvmOpcode::kRead);
+  queue_entry.fuse = 0b00;  // Normal operation, not fused
+  queue_entry.psdt = 0b00;  // PRPs are used for data transfer
+  queue_entry.mptr = reinterpret_cast<std::uint64_t>(AllocPages(1));
+
+  queue_entry.nsid = nsid;
+
+  queue_entry.dptr.prp.prp1 = reinterpret_cast<std::uint64_t>(AllocPages(1));
```

Comment:
Will need to define a way of constructing PRP lists as the read command is defined to use them in dptr field, per https://nvmexpress.org/wp-content/uploads/NVM-Express-1_4-2019.06.10-Ratified.pdf section 6.9

Reply:
Given recent developments in communication with the NVMe device this may not be nessesary